larq_zoo/literature/real_to_bin_nets.py

```diff
@@ -0,0 +1,491 @@
+        )(z)
+        dim_expansion = tf.keras.layers.Dense(
+            out_filters,
+            activation="sigmoid",
+            kernel_initializer="he_normal",
+            kernel_regularizer=self.kernel_regularizer,
+            name=f"{name}_scaling_dense_expand",
+            use_bias=False,
+        )(dim_reduction)
+        scales = tf.keras.layers.Reshape(
+            (1, 1, out_filters), name=f"{name}_scaling_reshape"
+        )(dim_expansion)
+
+        return tf.keras.layers.Multiply(name=f"{name}_scaling_multiplication")(
+            [conv_output, scales]
+        )
+
+    def _get_weights_path(self):
+        if (
+            not self.kernel_quantizer == "ste_sign"
+            and self.input_quantizer == "ste_sign"
+        ):
+            raise NotImplementedError(
+                f"{self.model_name} only has stored weights for the BNN variant"
+            )
+        if self.include_top:
+            weights_path = utils.download_pretrained_model(
+                model="r2b",
+                version="v0.1.0",
+                file="r2b_weights.h5",
+                file_hash="e8fd16ca1ab9810ac3835f24f5c62758a57bc32a615f73aaa50d382d2b9617e1",
+            )
+        else:
+            weights_path = utils.download_pretrained_model(
+                model="r2b",
+                version="v0.1.0",
+                file="r2b_weights_notop.h5",
+                file_hash="4ec47abf1a4da5c65f4908076257e8d5c812673891089a88c9d9e84e949d1dab",
+            )
+        return weights_path
+
+
+@factory
+class ResNet18Factory(_SharedBaseFactory):
+    """Constructor for a ResNet18 with layer names matching Real-to-Bin nets."""
+
+    def block(self, x, downsample=False, name=""):
+        """One full residual block, consisting of two convolutions.
+
+        This follows the definition of a "block" from Figure 1 (Left) of Martinez et al.
+        """
+
+        in_channels = x.shape[-1]
+        out_channels = in_channels * 2 if downsample else in_channels
+
+        # Shortcut, which gets downsampled if necessary
+        shortcut_add = self.shortcut_connection(x, name, in_channels, out_channels)
+
+        for convolution in ["a", "b"]:
+            x = tf.keras.layers.Conv2D(
+                filters=out_channels,
+                kernel_size=3,
+                kernel_regularizer=self.kernel_regularizer,
+                strides=2 if downsample and convolution == "a" else 1,
+                padding="same",
+                name=f"{name}{convolution}_conv2d",
+            )(x)
+            x = tf.keras.layers.BatchNormalization(
+                momentum=self.momentum, name=f"{name}{convolution}_batch_norm"
+            )(x)
+            x = tf.keras.layers.Activation("relu", name=f"{name}{convolution}_relu")(x)
+
+        x = tf.keras.layers.Add(name=f"{name}_skip_add")([x, shortcut_add])
+        return tf.keras.layers.ReLU(name=f"{name}_out")(x)
+
+
+@factory
+class StrongBaselineNetBAN(StrongBaselineNetFactory):
```

Comment:
Do you mind if we stick to appending `Factory` to the names of these (including the five or so directly below)? I know it makes the names quite long but I like the consistency.

Reply:
Implemented in d4b4c78